plugins/woocommerce/src/LayoutTemplates/LayoutTemplateRegistry.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+namespace Automattic\WooCommerce\LayoutTemplates;
+
+use Automattic\WooCommerce\Admin\BlockTemplates\BlockTemplateInterface;
+
+/**
+ * Layout template registry.
+ */
+final class LayoutTemplateRegistry {
```

Comment:
One thing this class might benefit from is a getter of some type that handles instantiating a template when first called.

If the templates have not been instantiated, we won't be able to take advantage of its methods, e.g., `get_area()`, `add_group()`, etc.

Reply:
`instantiate_layout_templates()` will instantiate templates and return them.

But, currently...

(1) It doesn't cache the instances, so if you call that method twice, new instances will be returned each time, which is not expected and confusing
(2) The method name doesn't clearly indicate that this is the method you can use to get a specific instance

I'll tweak things a bit to address these points!

Reply:
I've changed the implementation such that a template is only instantiated once.

In order to provide extensions with a robust way to ensure make template modifications via methods such as `get_area()`, `add_group()`, I've added two actions that they can hook into. This keeps the instantiation of templates in the control of the code registry implementation, so that extensions do not instantiate templates that will never end up being used for a given PHP request.

- `woocommerce_layout_template_before_instantiation`
- `woocommerce_layout_template_after_instantiation`

I think this strikes a good balance between flexibility and control.

Reply:
> Nice! I think this is a really good way of handling things. Just for the sake of discussion, how do you see the `before` hook being used?
> 
> I think there are probably good use cases, but just want to see what those are in case there's any chance we can limit down the number of introduced hooks.

The `before` hook should be used to add template modification hooks that run when blocks are added/removed from templates...

- `woocommerce_product_editor_block_template_{template_name}_after_add_block_{block_id}`
- `woocommerce_product_editor_block_template_after_add_block`
- `woocommerce_product_editor_block_template_{template_name}_after_remove_block_{block_id}`
- `woocommerce_product_editor_block_template_after_remove_block`

(from https://github.com/woocommerce/woocommerce/blob/trunk/docs/product-editor-development/block-template-lifecycle.md, which I realize I also need to update in this PR!)

Reply:
> Also as a side note, in general we have multiple ways of naming things:
> 
> * `noun_pre_action`
> * `noun_before_action`
> * `before_action_noun`
> 
> Personally, I like to go from least specific to most specific which would be `noun_action_before` but I realize that's not the most common pattern used and not overly opinionated on this, but long term it would be great to have a consistent pattern for this.

I agree that consistency would be ideal. I think I'll stick with this naming as it matches the pattern used for the actions referenced in my comment above.

Reply:
I really see the `before` hook being the preferred hook to use (with the actions referenced above), but recognize the `after` way of doing things can be easier to conceptually understand (though, there is more manual error checking that needs to be done if you take that approach, since you might be trying to get sections/groups/etc that have been removed). And, the `after` way is definitely easier if you want to add a new group that isn't dependent on anything else in the template.

Reply:
Correct me if I'm wrong, but can't the other before block hooks (e.g., `woocommerce_product_editor_block_template_{template_name}_after_add_block_{block_id}`) be used without the need to hook into `woocommerce_layout_template_before_instantiation` and save an extra step?  For example:

```php
add_action( 'woocommerce_product_editor_block_template_simple_product_after_add_block_general', 'add_my_block' );
```

If I'm correct in thinking that, I'm wondering how we envision the `woocommerce_layout_template_before_instantiation` hook specifically being used.  If we don't have any current cases, I'd be in favor of removing for now just to keep the public API coverage we need to support as small as possible.

Reply:
True, there is no need to use the `woocommerce_layout_template_before_instantiation`.

However, I think this hook is valuable if recommended as the way to hook up the block hooks, because:

- The dev knows that if they hook up their block hooks in here, they will be hooked up in time. Though, given that now the layout templates aren't instantiated until late in the REST API endpoint request lifecycle, it is unlikely they would hook them up afterwards.
- The hooks will only be hooked up when the layout templates REST API endpoint requests are being processed

We could leave it out for now though, if you think this is premature optimization. I suppose hooking up hooks isn't expensive... I just shudder when I think about how much code overall ends up getting executed that isn't relevant for most WP/WC pages, so I try to avoid that when I can.